src/sinks/kafka.rs

```diff
@@ -141,95 +151,136 @@ impl KafkaSink {
     fn new(config: KafkaSinkConfig, acker: Acker) -> crate::Result<Self> {
         let producer = config.to_rdkafka()?.create().context(KafkaCreateFailed)?;
         Ok(KafkaSink {
-            producer,
+            producer: Arc::new(producer),
             topic: Template::try_from(config.topic).context(TopicTemplate)?,
             key_field: config.key_field,
             encoding: config.encoding.into(),
+            delivery_fut: FuturesUnordered::new(),
             in_flight: FuturesUnordered::new(),
             acker,
             seq_head: 0,
             seq_tail: 0,
             pending_acks: HashSet::new(),
         })
     }
+
+    fn poll_delivery_fut(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        while !self.delivery_fut.is_empty() {
+            let result = Pin::new(&mut self.delivery_fut).poll_next(cx);
+            let (seqno, result) = ready!(result).expect("`delivery_fut` is endless stream");
+            self.in_flight.push(Box::pin(async move {
+                let result = match result {
+                    Ok(fut) => {
+                        fut.map_ok(|result| result.map_err(|(error, _owned_message)| error))
+                            .await
+                    }
+                    Err(error) => Ok(Err(error)),
+                };
+
+                (seqno, result)
+            }));
+        }
+
+        Poll::Ready(())
+    }
 }
 
-impl Sink for KafkaSink {
-    type SinkItem = Event;
-    type SinkError = ();
+impl Sink<Event> for KafkaSink {
+    type Error = ();
+
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if matches!(self.poll_delivery_fut(cx), Poll::Pending)
+            && self.delivery_fut.len() >= SEND_RESULT_LIMIT
+        {
+            Poll::Pending
+        } else {
+            Poll::Ready(Ok(()))
+        }
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: Event) -> Result<(), Self::Error> {
+        assert!(
+            self.delivery_fut.len() < SEND_RESULT_LIMIT,
+            "Expected to `poll_ready` called first."
+        );
 
-    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
         let topic = self.topic.render_string(&item).map_err(|missing_keys| {
             error!(message = "Missing keys for topic.", missing_keys = ?missing_keys);
         })?;
-
         let (key, body) = encode_event(item.clone(), &self.key_field, &self.encoding);
 
-        let mut record = FutureRecord::to(&topic).key(&key).payload(&body[..]);
+        let seqno = self.seq_head;
+        self.seq_head += 1;
 
-        if let Some(Value::Timestamp(timestamp)) = item.as_log().get(log_schema().timestamp_key()) {
-            record = record.timestamp(timestamp.timestamp_millis());
-        }
+        let producer = Arc::clone(&self.producer);
+        self.delivery_fut.push(Box::pin(async move {
```

Comment:
`send_result` is not async and previously we returned `NotReady` if the queue was full. In the new `Sink` trait we can not return not ready, so for the case when sending blocked (queue full) I added `FuturesUnordered` where `send_result` called with 10ms delay. A number of futures limited by `SEND_RESULT_LIMIT` (set to 5).
I'm not completely sure that this is the best solution, maybe have 1 future with 1ms (or something like this) delay will be better.

Reply:
Just to follow up on this, I think what you have here is reasonable.

If we wanted to be more precise, we could try to tie the retries to the completion of the `in_flight` futures instead of a delay. So when one of those is complete we try again to send data, assuming that the completed request could have opened up space in the buffer. But that could be a bit tricky.

Reply:
Oh, right, that sounds more correct way. I agree that this can be tricky, but should I try to add it?

Reply:
It's probably worth spending a little bit of time to see how it would work, but not a lot if it turns out to be complicated.

Reply:
Added in https://github.com/timberio/vector/pull/5007